### PR TITLE
feat: add support for multi-threading flint

### PR DESCRIPTION
**Pull request by lucifer1004**
Solves #131 

The code was copied from `Nemo.jl` with minor modifications (replacing `Nemo` with `Arblib`). The usage has been added to `README.md`.

### CODE REVIEW

**Comment by lucifer1004** [comment]
It seems that we have to use `NEMO_THREADED` since it was hard-coded [here](https://github.com/JuliaBinaryWrappers/FLINT_jll.jl/blob/9a4e5a48246b4a7c4048ff1d2d574a30a972d705/src/wrappers/x86_64-linux-gnu.jl#L17).

File: src/Arblib.jl [hunk 1]
```diff
@@ -41,6 +42,25 @@ macro libflint(function_name)
...
+function __init__()
+    __isthreaded[] = Base.get(ENV, "NEMO_THREADED", "") == "1"
+
+    if __isthreaded[]
+        ccall((:__gmp_set_memory_functions, :libgmp), Nothing,
+            (Int, Int, Int), 0, 0, 0)
+    end
+end
+
+function flint_set_num_threads(a::Int)
```
Comment (kalmarek) [inline src/Arblib.jl:56]:
```suggestion
function flint_set_num_threads(a::Integer)
```
anything that `cconverts` to `Clong` should be fine

File: src/Arblib.jl [hunk 2]
```diff
@@ -41,6 +42,25 @@ macro libflint(function_name)
...
+    if !__isthreaded[]
+        error("To use threaded flint, julia has to be started with NEMO_THREADED=1")
+    else
+        ccall(@libflint(flint_set_num_threads), Nothing, (Int,), a)
```
Comment (kalmarek) [inline src/Arblib.jl:60]: is that a `Clong`, or `Cint` in flint?

File: src/Arblib.jl [hunk 3]
```diff
@@ -41,6 +42,25 @@ macro libflint(function_name)
     return (:($function_name), Arb_jll.libflint)
 end
 
+const __isthreaded = Ref(false)
+
+function __init__()
+    __isthreaded[] = Base.get(ENV, "NEMO_THREADED", "") == "1"
+
+    if __isthreaded[]
+        ccall((:__gmp_set_memory_functions, :libgmp), Nothing,
```
Comment (kalmarek) [inline src/Arblib.jl:51]: why do we need to set those memory functions? also is that really `0::Int` or rather `C_NULL::Ptr{Cvoid}`?

**Review by kalmarek — CHANGES_REQUESTED** [review]

File: src/Arblib.jl (hunk 3 above)
Comment (lucifer1004) [reply src/Arblib.jl:51]: I just copied the codes from `Nemo.jl` codebase, since I am not familiar with `gmp` or `flint`.

File: src/Arblib.jl (hunk 2 above)
Comment (kalmarek) [reply src/Arblib.jl:60]: It's actually `Cint`: https://flintlib.org/doc/flint.html?highlight=flint_set_num_threads#c.flint_set_num_threads
```suggestion
        ccall(@libflint(flint_set_num_threads), Nothing, (Cint,), a)
```

File: src/Arblib.jl (hunk 3 above)
Comment (kalmarek) [reply src/Arblib.jl:51]: @Joel-Dahne could you chip in?

**Comment by Joel-Dahne** [comment]
Thanks for taking a look at this, it would be great to get working! Let me take a look at it tomorrow!

File: src/Arblib.jl (hunk 3 above)
Comment (Joel-Dahne) [reply src/Arblib.jl:51]: It works for me even if I don't set these memory functions, but I don't know what they do so it could depend on your system or on what you are doing for example. It actually works even without setting the `NEMO_THREADED` to `1` for me, so I don't even know what that part does...

It could maybe be that Nemo has these methods for historical reasons, they use Flint before `flint_jll` was a thing, and that they are not needed anymore.

How about we try to just not set these memory functions for now and hope that things work? And then we add it back if we notice any issues? I would like to avoid touching gmp things which I don't understand if we can avoid it.

Comment (lucifer1004) [reply src/Arblib.jl:51]: Are you suggesting we just wrap the `flint_set_num_threads` function and do nothing else?

File: README.md [hunk 4]
```diff
@@ -29,18 +29,18 @@ base Julia. They contain a pointer to an object of the corresponding
...
+| `acb_t`      | `Acb` or `AcbRef`             |
+| `arb_t*`     | `ArbVector` or `ArbRefVector` |
+| `acb_t* `    | `AcbVector` or `AcbRefVector` |
 | `arb_mat_t`  | `ArbMatrix` or `ArbRefMatrix` |
-| `acb_mat_t`  | `AcbMatrix` or `AcbRefMatrix`  |
-| `arb_poly_t`  | `ArbPoly` or `ArbSeries` |
-| `acb_poly_t`  | `AcbPoly` or `AcbSeries` |
+| `acb_mat_t`  | `AcbMatrix` or `AcbRefMatrix` |
+| `arb_poly_t` | `ArbPoly` or `ArbSeries`      |
+| `acb_poly_t` | `AcbPoly` or `AcbSeries`      |
```
Comment (Joel-Dahne) [inline README.md:43]: Would it be possible to revert this change? Since it's not part of the threading.

File: src/Arblib.jl [hunk 5]
```diff
@@ -5,10 +5,10 @@ import LinearAlgebra
 import SpecialFunctions
 
 if VERSION >= v"1.5.0-DEV.639"
-    import Base: contains
+    import Base:contains
 end
 if VERSION >= v"1.6"
-    import Base: sincospi
+    import Base:sincospi
```
Comment (Joel-Dahne) [inline src/Arblib.jl:11]: Would it be possible to reverse this as well? Again not part of the treading.

**Review by Joel-Dahne — COMMENTED** [review]
Other than the comments above this looks good and seems to work well on my machine as well! I think it would be a good idea to add some tests for it as well. I would propose adding this
```julia
@testset "Threading" begin
    @testset "Check number of threads" begin
        @test ccall(Arblib.@libflint(flint_get_num_threads), Cint, ()) == 1

        Arblib.flint_set_num_threads(2)
        @test ccall(Arblib.@libflint(flint_get_num_threads), Cint, ()) == 2

        Arblib.flint_set_num_threads(4)
        @test ccall(Arblib.@libflint(flint_get_num_threads), Cint, ()) == 4

        Arblib.flint_set_num_threads(1)
        @test ccall(Arblib.@libflint(flint_get_num_threads), Cint, ()) == 1
    end

    @testset "Disabled threading" begin
        # This manually unsets the threading flag and checks that we
        # get an error when trying to set the number of threads
        Arblib.__isthreaded[] = false

        @test_throws ErrorException Arblib.flint_set_num_threads(2)

        Arblib.__isthreaded[] = true
    end

    @testset "Matrix multiplication" begin
        A = ArbMatrix(reshape(1:10000, 100, 100))

        Arblib.flint_set_num_threads(1)
        B1 = Arblib.mul_threaded!(similar(A), A, A)

        Arblib.flint_set_num_threads(2)
        B2 = Arblib.mul_threaded!(similar(A), A, A)

        @test B1 == B2 == A * A

        Arblib.flint_set_num_threads(2)
    end
end
```
in a file `test/threading.jl` and then updating `test/runtest.jl` to be
```julia
ENV["NEMO_THREADED"] = 1

using Arblib, Test, LinearAlgebra, SpecialFunctions

@testset "Arblib" begin
    include("arb_types-test.jl")
    include("types-test.jl")
    include("arbcall-test.jl")
    include("precision-test.jl")
    include("setters-test.jl")
    include("constructors-test.jl")
    include("predicates-test.jl")
    include("show-test.jl")
    include("promotion.jl")
    include("examples.jl")
    include("arithmetic.jl")
    include("rand.jl")
    include("float.jl")
    include("interval.jl")
    include("vector.jl")
    include("matrix.jl")
    include("eigen.jl")
    include("calc_integrate.jl")
    include("ref-test.jl")
    include("poly.jl")
    include("series.jl")
    include("special-functions.jl")
    include("threading.jl")
end
```

File: src/Arblib.jl (hunk 3 above)
Comment (Joel-Dahne) [reply src/Arblib.jl:51]: Exactly! At least it works on my Linux machine. I don't know what you run on but it would be good if you could test that it works as well. Also if we add the tests I propose below they will also be run on both Windows and Mac with the continuous integration.

**Comment by Joel-Dahne** [comment]
Oh, and it seems like the formatter was complaining as well, you can format the code with
```julia
using JuliaFormatter
format(".")
```

**Comment by lucifer1004** [comment]
@Joel-Dahne @kalmarek Please review the latest commit.

**Comment by Joel-Dahne** [comment]
This looks good to me! Great job! Lets see if @kalmarek has anything to add as well. Normally I would propose to push the version as well, to make a new release. However I plan to push some other changes later today or tomorrow, so we can maybe wait with the new release until then.

File: src/Arblib.jl [hunk 6]
```diff
@@ -31,7 +31,8 @@ export Mag,
     ArbSeries,
     AcbPoly,
     AcbSeries,
-    ref
+    ref,
+    flint_set_num_threads
```
Comment (kalmarek) [inline src/Arblib.jl:35]: we shouldn't export this function

**Review by kalmarek — COMMENTED** [review]
other than exporing `flint_set_num_threads` this looks good!

**Comment by kalmarek** [comment]
on a slight tangent here: @lucifer1004 remember that Nemo.jl is GPL licensed while Arblib.jl is MIT;
copying such small snippet shouldn't be a problem, but in general it's advisable not to look at GPL code when submitting to MIT licensed project.

File: src/Arblib.jl (hunk 6 above)
Comment (lucifer1004) [reply src/Arblib.jl]: Updated

**Review by kalmarek — APPROVED** [review]

**Comment by kalmarek** [comment]
@lucifer1004 Thanks for contributing!
